Reject running `compiletest` self-tests against stage 0 rustc unless explicitly allowed

jieyouxu opened the pull request:
Currently, in `pr-check-1`, we run `python3 ../x.py test --stage 0 src/tools/compiletest`, which is `compiletest` self-tests against stage 0 rustc. This makes it very annoying for PRs wanting to change target spec JSON format, which `compiletest` depends on for target information, as otherwise `compiletest` would have to know how to handle 2 different target spec JSON formats and know when to pick which.

Instead of doing that, we change `compiletest` self-tests to reject running against stage 0 `rustc` *unless* explicitly allowed with `build.compiletest-allow-stage0=true`. `build.compiletest-allow-stage0` is a proper bootstrap config option in favor of the ad-hoc `COMPILETEST_FORCE_STAGE0` env var. This means that:

- `./x test src/tools/compiletest --stage=0` is not allowed, unless `build.compiletest-allow-stage0=true` is set. In this scenario, `compiletest` self-tests should be expected to fail unless the stage 0 `rustc` as provided is like codegen_cranelift where it's *actually* built from in-tree `rustc` sources.
- In CI, we change `./x test src/tools/compiletest --stage=0` to `./x test src/tools/compiletest --stage=1`, and move it to `pr-check-2`. Yes, this involves building the stage 1 compiler, but `pr-check-2` already has to build stage 1 compiler to test stage 1 library crates.
- Crucially, this means that **`compiletest` is only intended to support one target spec JSON format**, namely the one corresponding to the in-tree `rustc`.
- This should preserve the `compiletest-use-stage0-libtest` UX optimization where changing `compiler/` tree should still not require rebuilding `compiletest` as long as `build.compiletest-use-stage0-libtest=true`, as that should remain orthogonal. This is completely unlike my previous attempt at https://github.com/rust-lang/rust/pull/144563 that tries to do a way more invasive change which would cause the rebuild problem.

Best reviewed commit-by-commit.

---

r? @Kobzol

jieyouxu committed:
Run `compiletest` self-tests against stage 1 rustc

And move `./x test compiletest --stage=1` to `pr-check-2`, where testing
library artifacts already requires building the stage 1 compiler.

jieyouxu committed:
Add change tracker entry

jieyouxu committed:
Update `codegen_{cranelift,gcc}` and `opt-dist` to use `build.compiletest-allow-stage0`

Kobzol reviewed (commented):
`opt-dist` and `cg_gcc` will also need to be updated (you can grep for usage of `COMPILETEST_FORCE_STAGE0`).

jieyouxu commented on bootstrap.example.toml:
```diff
@@ -465,6 +465,12 @@
+# suites to be run against the stage0 rustc. Note that this configuration is
+# unsupported and tests will almost always fail, and is only intended to be used
+# by e.g. `codegen_cranelift` via `local-rebuild`.
+#build.compiletest-force-stage0 = false
```
I think I prefer "force" in that it overrides a warning about you doing a unsupported-outside-of-a-niche-use-case thing. But I don't feel *too* strongly about the naming.

Kobzol replied:
It is a bit misleading w.r.t. the comment though. 

> Whether to force `compiletest` self-tests and `compiletest`-managed test
> # suites to be run against the stage 0 rustc.

This claims that if you set it to `true`, the test suites will be force to run against stage 0 rustc, which is not true.

jieyouxu replied:
Changed from "force" -> "allow"

jieyouxu committed:
Introduce proper `build.compiletest-allow-stage0` config option

In favor of the adhoc `COMPILETEST_FORCE_STAGE0` env var.

jieyouxu committed:
Deny `compiletest` self-tests being run against stage 0 rustc unless explicitly allowed

Otherwise, `compiletest` would have to know e.g. how to parse two
different target spec, if target spec format was changed between beta
`rustc` and in-tree `rustc`.

Kobzol commented:
Thanks!

@bors r+
